app/lib/search/updater.dart

```diff
@@ -4,92 +4,107 @@
 
-  Future _doUpdate(int limit) async {
-    final List<Key> versionKeys = [];
-    final Map<String, String> devVersions = {};
+class BatchIndexUpdater {
+  final List<Task> _batch = [];
+  Timer _batchUpdateTimer;
+  Future _ongoingBatchUpdate;
+  int _taskCount = 0;
 
-    _logger.info('Polling packages for changes.');
-    int count = 0;
-    await for (Package p in dbService.query(Package).run()) {
-      final bool inIndex = await index.contains(
-          _toUrl(p.name), p.latestVersion, p.latestDevVersion);
-      if (inIndex) continue;
-      versionKeys.add(p.latestVersionKey);
-      devVersions[p.name] = p.latestDevVersion;
-      count++;
-      if (limit != null && limit <= count) break;
+  Future updateIndex(Task task) async {
+    await _currentBatchRunning();
+    _batch.add(task);
+    _taskCount++;
+    if (_batch.length < 20) {
+      _batchUpdateTimer ??= new Timer(const Duration(seconds: 10), () {
+        _updateBatch();
+      });
+    } else {
+      await _updateBatch();
     }
-    _logger.info('Found ${versionKeys.length} packages to update.');
-    if (versionKeys.isEmpty) return;
+  }
 
-    for (List<Key> slices in sliceList(versionKeys, 20)) {
-      final Key firstKey = slices.first;
-      final String firstPkg = firstKey.parent.id;
-      final String firstVersion = firstKey.id;
-      _logger.info('Updating packages staring with $firstPkg $firstVersion');
-      final List<PackageVersion> versions = await dbService.lookup(slices);
-      final List<PackageDocument> documents = versions
-          .map((pv) => new PackageDocument(
-                url: _toUrl(pv.package),
-                package: pv.package,
-                version: pv.version,
-                devVersion: devVersions[pv.package],
-                detectedTypes: pv.detectedTypes,
-                description: pv.pubspec.description,
-                lastUpdated: pv.shortCreated,
-                readme: pv.readmeContent,
-                popularity: mockScores[pv.package] ?? 0.0,
-              ))
-          .toList();
-      await index.addAll(documents);
-      _logger.info('Updated ${slices.length} packages.');
-    }
+  Future _updateBatch() async {
+    _batchUpdateTimer?.cancel();
+    _batchUpdateTimer = null;
 
-    _logger.info('Calling index.merge() ...');
-    await index.merge();
-    _logger.info('index.merge() completed.');
+    await _currentBatchRunning();
+    if (_batch.isEmpty) return;
+
+    final Completer completer = new Completer();
+    _ongoingBatchUpdate = completer.future;
+
+    try {
+      final List<Task> tasks = new List.from(_batch);
+      _batch.clear();
+      _logger.info(
+          'Updating index with ${tasks.length} packages [example: ${tasks.first
+              .package}]');
+      final List<PackageDocument> docs = await searchBackend
+          .loadDocuments(tasks.map((t) => t.package).toList());
+      await packageIndex.addAll(docs);
+      final bool doMerge = _firstPackageListCount != null &&
```

Comment:
If you need shared state, instead of using a global variable, you could consider just letting one class know the other:
```dart
class BatchIndexUpdater {
  final IndexUpdateTaskSource source;

  Future _updateBatch() async {
    ...
    if (source.firstScanComplete) {
      ...
    }
  }
}

```